Update stokes most

gustavo-marques opened the pull request:
* Add new subroutine `cvmix_kpp_compute_ER_depth` to calculate the entrainment flux and corresponding depth using the entrainment rule. For each assumed boundary layer depth (`OBL_depth`) at cell centers, the subroutine iteratively searches for the first depth where the entrainment depth (`ERdepth`) becomes positive. The corresponding index (`kER_depth`) is also returned. If no viable solution is found, the subroutine returns `ERdepth = -1` and `kER_depth = -1`.
* Add new subroutine `cvmix_kpp_quad_fit` to estimate the maximum of a quadratic fit. 
* Update `cvmix_kpp_compute_StokesXi`
* Compute Shear, Stokes, Buoyancy, surface-layer TKE Production

gustavo-marques committed:
Update stokes_most

gustavo-marques committed:
Delete unsed argument, deltaRho

gustavo-marques committed:
Replace *SLD with *SL

gustavo-marques committed:
Replace uSbar_SL => uSb_SL; vSbar_SL => vSb_SL

gustavo-marques committed:
Fix args passed to cvmix_kpp_driver

gustavo-marques committed:
Set SLDepth=cvmix_one and change BLDepth -> BLdepth

mnlevy1981 commented on src/shared/cvmix_kpp.F90:
```diff
@@ -88,7 +88,8 @@ module cvmix_kpp
   public :: cvmix_kpp_ustokes_SL_model
-
+  public :: cvmix_kpp_compute_ER_depth
+  public :: cvmix_kpp_quad_fit
```
There's a polynomial fit in [`cvmix_math.F90`](https://github.com/CVMix/CVMix-src/blob/master/src/shared/cvmix_math.F90#L59), though it needs a little updating because the wants two points and a derivative instead of the three points used here. Still, it makes more sense to add this functionality to the existing routine rather than create a new subroutine here.

mnlevy1981 commented on src/shared/cvmix_kpp.F90:
```diff
@@ -3374,78 +3375,75 @@ end subroutine cvmix_kpp_composite_Gshape
+    Cs = 0.038_cvmix_r8
+    Cb = 0.96_cvmix_r8
+    CempCGm = 3.5_cvmix_r8
+    CempCGs = 4.7_cvmix_r8
```
These should be declared as `real(cvmix_r8), parameter` (and it would be nice to see a comment describing units / where these values come from)

gustavo-marques replied:
Done!

mnlevy1981 commented on src/shared/cvmix_kpp.F90:
```diff
@@ -3374,78 +3375,75 @@ end subroutine cvmix_kpp_composite_Gshape
     ! Parameterized Buoyancy production of TKE
-    PBfact =  0.110_cvmix_r8
-    PB      = PBfact * MAX( -surf_buoy_force * BLdepth ,  cvmix_zero )
+    PBfact = 0.0893759_cvmix_r8
+    PB     = MAX( -surfBuoy_NS * BLdepth * PBfact , cvmix_zero )
+    PBfact = 0.00215_cvmix_r8 * CempCGs
+    PB     = PB + PBfact * BLdepth * ( abs(surf_buoy_force) - surf_buoy_force )
```
I think `PBfact` could also be a Fortran parameter (`0.0893759_cvmix_r8`) and then

```fortran
PB = MAX( -surfBuoy_NS * BLdepth * PBfact , cvmix_zero ) &
   + PBfact * CempCGs * BLdepth * ( abs(surf_buoy_force) - surf_buoy_force )
```

gustavo-marques replied:
PBfact cannot be used as a parameter because it is used with different values.

mnlevy1981 commented on src/shared/cvmix_kpp.F90:
```diff
@@ -3490,22 +3489,189 @@ subroutine cvmix_kpp_compute_StokesXi (zi, zk, kSL, SLDepth,          &
+
+  nlev      = size(OBL_depth)
+  Cemp_Rs   = 4.7_cvmix_r8
+!  Gat1      = cvmix_zero
```
We can just remove lines instead of commenting them out

gustavo-marques replied:
Done!

mnlevy1981 commented on src/shared/cvmix_kpp.F90:
```diff
@@ -3490,22 +3489,189 @@ subroutine cvmix_kpp_compute_StokesXi (zi, zk, kSL, SLDepth,          &
+  Cemp_Rs   = 4.7_cvmix_r8
+!  Gat1      = cvmix_zero
+  Fdelrho   = cvmix_one
+! kinv      = MAXLOC( Nsq )      ! interface index of maximum stratification, N2>0 (inversion)
+! maxNsq    = Nsq( kinv )
```
We can just remove lines instead of commenting them out

gustavo-marques replied:
Done!

mnlevy1981 commented on src/shared/cvmix_kpp.F90:
```diff
@@ -3490,22 +3489,189 @@ subroutine cvmix_kpp_compute_StokesXi (zi, zk, kSL, SLDepth,          &
+  sigma(:)  = cvmix_zero
+  Bflux(1)  = Bsfc_ns    ! non-solar surface buoyancy boundary condition for all kbl
+!                             Set OBL_depth(1)=top cell center values
+  BEnt(1)   = cvmix_zero
```
This is a duplicate of line 3591; should it be removed, or should it be `BEnt(2)` instead of `BEnt(1)`? Looking at how `BEnt` is used, I see a couple of `BEnt(kbl+1)` but no other index is evaluated. Maybe this should be a scalar instead of an array?

gustavo-marques replied:
`BEnt` is now a scalar.

mnlevy1981 reviewed (changes requested):
Sorry to sit on this for a month, but there are some minor updates needed. I'm happy to help you with them next week.

gustavo-marques committed:
Addresses Mike Levy comments

gustavo-marques commented:
Thank you, @mnlevy1981, for all the suggestions. We have addressed all comments except for switching to the polynomial fit in [cvmix_math.F90](https://github.com/CVMix/CVMix-src/blob/master/src/shared/cvmix_math.F90#L59).

mnlevy1981 committed:
Remove cvmix_kpp_quad_fit

The function cvmix_kpp_quad_fit did two things:

1. Given three values of (sigma, Bflux), find quadratic polynomial that
   interpolates function
2. Find the root of the derivative of the quadratic, and evaluate the
   polynomial at that point

I created an interface for cvmix_math_poly_interp() to do the interpolation
(that routine could do quadratic interpolation, but based on two points and a
derivative instead of three points), and then find the derivative and evaluate
the polynomial in cvmix_kpp_compute_ER_depth() instead of a separate function

mnlevy1981 committed:
Fix whitespace

gustavo-marques commented:
@mnlevy1981, here is the difference in SST when adding your polynomial fit after 1 month ( B = baseline):

<img width="931" height="1124" alt="image" src="https://github.com/user-attachments/assets/cdf77e04-067f-4805-8dc4-06e92a75f2e6" />

This is more than I expected, so I will run for longer to check the effects on climate-related metrics.

gustavo-marques commented:
This PR has minimal impact on climate metrics. 
I approve it.
